### PR TITLE
Turn named return parameters in function types into an error.

### LINKED ISSUE #2428: Should function types be allowed to name their arguments?
The example for function types in the documentation looks like this, where function parameters can have names. 

```javascript
  // ...
  function reduce(
    uint[] memory self,
    function (uint x, uint y) returns (uint) f
  )
    internal
    returns (uint r)
  {
    r = self[0];
    for (uint i = 1; i < self.length; i++) {
      r = f(r, self[i]);
    }
  }
  // ...
```

This behavior was explicitly forbidden in the grammar specification (see [here](https://github.com/ethereum/solidity/blob/develop/docs/grammar.txt#L50-L51)). Is this a bug in the current implementation?

### PR DESCRIPTION
**Pull request by ekpyron**
Closes  #2428.

### CODE REVIEW

**Comment by ekpyron** [comment]
Alternatively, the parser could be changed and already disallow named parameters in function types.

**Comment by axic** [comment]
I think it is better this way, since the parser's error messages aren't that great.

**Review by axic — DISMISSED** [review]

**Comment by axic** [comment]
Wait a second, I think we agreed to leave the named arguments: https://github.com/ethereum/solidity/issues/2428#issuecomment-333175533

But the return parameters should be changed.

**Comment by ekpyron** [comment]
@axic Since both cases were deprecated in #2464, I think it makes sense to turn both into an error now.

**Comment by axic** [comment]
If you look at the timestamp, the comment I've linked came after the deprecation.

**Comment by ekpyron** [comment]
@axic @chriseth I forgot about this during our call - I think it makes sense to remove named arguments in function types altogether (I think that overwriting the names during assignment as well as keeping them will both be weird, so best not to have them at all), but if you disagree I can restrict this PR to only the return parameters.

**Comment by ekpyron** [comment]
Discussion on gitter:

@chriseth:
I'm undecided about the named parameters in function types

@axic:
I am on the opinion that calling with name parameters is very useful, and should push in that direction. It really makes sense when doing for example ERC20 approvals as it is clear what each argument means.

@ekpyron:
Yes, naming parameters is nice... but to me both disallowing assigning a function type variable to a function with arguments that have different names, as well as allowing it (which parameter names does the type have after assignment?) feels strange...
To me it feels more natural not to consider parameter names part of the function signature, resp. the function type...
I would tend to remove it for now (it was deprecated already and calling with named arguments seems to never have worked for function types) - then later we can consider adding it back with proper and working behavior.
But if you disagree I'm also fine with removing that part from #4219 for now.

@axic
Good point about functions assigned to functiontypes with different names. I am not sure, we do allow overriding in inheritance with different names, I consider that the same.

**Comment by ekpyron** [comment]
So should the following be allowed then?

```
contract C {
   function f(address recipient, address sender) { ... }
   function g() {
     function(address sender, address recipient) localf = f;
     localf({sender: ..., recipient: ...});
   }
}
```

**Comment by axic** [comment]
I would say yes, because it falls into the same category as:
```
interface A {
  function transfer(address sender, address recipient);
}
contract B is A {
  function transfer(address recipient, address sender) {}
}
```

**Comment by ekpyron** [comment]
OK, I'm still not convinced, but I restricted this PR to return parameters only for now, so that it can be merged.
I'm opening a new issue for looking into named arguments in function types, though, since at the moment they behave very differently from how @axic seems to want them to behave.

File: libsolidity/analysis/SyntaxChecker.cpp [hunk 1]
```diff
@@ -255,7 +255,7 @@ bool SyntaxChecker::visit(FunctionTypeName const& _node)
 
 	for (auto const& decl: _node.returnParameterTypeList()->parameters())
 		if (!decl->name().empty())
-			m_errorReporter.warning(decl->location(), "Naming function type return parameters is deprecated.");
+			m_errorReporter.syntaxError(decl->location(), "Return parameters in function types may not be named.");
```
Comment (axic) [inline libsolidity/analysis/SyntaxChecker.cpp:258]: I'm fine with either text, though we kept the "is deprecated" error for most of the other features. I think it makes sense for the long standing features, but this probably wasn't used at all so it shouldn't make a difference.

File: test/libsolidity/syntaxTests/parsing/function_type_in_struct.sol [hunk 2]
```diff
@@ -8,4 +8,4 @@ contract test {
 // ----
 // Warning: (49-55): Naming function type parameters is deprecated.
 // Warning: (57-63): Naming function type parameters is deprecated.
-// Warning: (83-89): Naming function type return parameters is deprecated.
+// SyntaxError: (83-89): Return parameters in function types may not be named.
```
Comment (axic) [inline test/libsolidity/syntaxTests/parsing/function_type_in_struct.sol]: I think this test case should be updated to remove the name and to remove the error. The error is tested for in the other test case.

**Review by axic — DISMISSED** [review]
Please adjust the test case otherwise I am fine with this.

**Comment by axic** [comment]
> OK, I'm still not convinced, but I restricted this PR to return parameters only for now, so that it can be merged.

I think it is worth discussing this a bit more and therefore agree with merging this PR in this restricted manner.

**Review by axic — APPROVED** [review]

**Comment by axic** [comment]
Okay I'm pulling the plug here now 😉

**Pull request merged**